In [4]:
# Import necessary libraries
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
                         summarize,
                         poly)
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.anova import anova_lm
import matplotlib.pyplot as plt

Question 1

In [12]:
## 1.1
data = pd.read_csv(r"D:\william\OneDrive - UW-Madison\UW-Madison\722\-2025AAE722_William-J\Advertising.csv")

X = pd.DataFrame({'intercept': np.ones(data.shape[0]),
                  'TV': data['TV'],
                  'radio': data['radio'],
                  'newspaper': data['newspaper']})
y = data['sales']

model = sm.OLS(y, X)
results1 = model.fit()
summarize(results1)

coef  std err       t  P>|t|
intercept  2.9389    0.312   9.422   0.00
TV         0.0458    0.001  32.809   0.00
radio      0.1885    0.009  21.893   0.00
newspaper -0.0010    0.006  -0.177   0.86

1.2

Te P-values of TV and Radio are both 0, which is less than the significance level of 0.05, so we reject the null hypothesis and conclude that there is a significant relationship between the advertising budgets for TV and Radio and the sales.
The P-value of Newspaper is 0.8599, which is greater than the significance level.
The 
The null hypothesis of these 3 predictors means that whether the coefficients of the advertising budgets for TV, Radio, and Newspaper are equal to 0. It helps us to identify if the relationship between advertising budgets and sales is statistically significant.

1.3

The results of the regression model show that 1000 dollars increase in the TV advertising budget is associated with an average increase of approximately 46 units in sales, holding all other variables constant. Similarly, a 1000 dollars increase in the Radio advertising budget is associated with an average increase of approximately 189 units in sales, holding all other variables constant. However, the Newspaper advertising budget does not have a statistically significant effect on sales, as indicated by its high p-value (0.8599). This suggests that changes in the Newspaper advertising budget do not have a meaningful impact on sales when controlling for TV and Radio advertising budgets.

Question 2

The KNN Classifier is used to classify the test data points by finding the 'k' nearest neighbors from the training data based on a distance metric and assigning the most common class among those neighbors to build a boundary then apply the boundary to classify the test data points.

The KNN Regression method is used to predict an outcome variable for a test data point by finding the 'k' nearest neighbors from the training data based on a distance metric and averaging the outcome variable of those neighbors to make the prediction.

The difference between KNN Classifier and KNN Regression is that KNN Classifier is used for categorical outcome variables, while KNN Regression is used for continuous outcome variables.

Question 3

a\)

I would expect the training RSS for the cubic regression to be lower than or equal to the training RSS for the linear regression.

Because the cubic model is more flexible and can fit the data at least as well as the linear model.


b\)

I would expect the test RSS for the cubic regression to be greater than or equal to the test RSS for the linear regression. 

Because if the true relationship between X and Y is linear, the linear regression model is correctly specified, while the cubic model is more likely to overfit and perform worse on unseen data.

c\)

If the true relationship between X and Y is not linear, I would expect the training RSS for the cubic regression to be lower than or equal to the training RSS for the linear regression.

Because the cubic model is more flexible and can fit the data at least as well as the linear model.

d\)

If the true relationship between X and Y is not linear, I think there is not enough information to tell which model will have a lower test RSS. It depends on how well the cubic model matches the true relationship and how much overfitting occurs.


Question 4

In [5]:
rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

In [6]:
# a)
model = sm.OLS(y, x)
resultsa = model.fit()
summarize(resultsa)

coef  std err       t  P>|t|
x1  1.9762    0.117  16.898    0.0

The estimated $\beta_1$ is 1.9762, the SE is 0.117, the t-statistic is 16.898, and the p-value is 0.000.

The model correctly identifies the relationship between X and Y, as the true relationship is linear with $\beta_1$ equal to 2. The estimated coefficient is close to the true value, and the p-value indicates that the relationship is statistically significant (reject the null hypothesis).

In [7]:
# b)
model = sm.OLS(x, y)
resultsb = model.fit()
summarize(resultsb)

coef  std err       t  P>|t|
x1  0.3757    0.022  16.898    0.0

The estimated $\beta_1$ is 0.376, the SE is 0.022, the t-statistic is 16.898, and the p-value is 0.000.

This result means that the coefficients of regression model $x$ onto $y$ and regression model $y$ onto $x$ are not simply reciprocals. The t-statistic is large and the p-value is very small, indicating that the relationship is statistically significant (reject the null hypothesis).

c\)

The results of (a) and (b) the slope from regressing y on x is not simply the reciprocal of the slope from regressing x on y.

It shows that the two regression slopes are not reciprocals unless the data are perfectly correlated and have equal variances.


d\)

SE($\hat{\beta}$) = $\frac{\sqrt{\sum(y_i^2-2x_iy_i\hat{\beta}+\hat{\beta}^2x_i^2)}}{\sqrt{(n-1)\sum(x_i^2)}}$ = $\frac{\sqrt{\sum(y_i^2-2\frac{(x_iy_i)^2}{x_i^2}+\frac{(x_iy_i)^2}{x_i^2})}}{\sqrt{(n-1)\sum(x_i^2)}}$ 

= $\frac{\sqrt{\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}{\sqrt{(n-1)\sum(x_i^2)}}$

$\frac{\hat{\beta}}{SE(\hat{\beta})}$ = $\frac{\sum(x_iy_i)^2}{\sum(x_i^2)}$ $\frac{\sqrt{(n-1)\sum(x_i^2)}}{\sqrt{\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}$ = $\frac{\sqrt{(n-1)}\sum(x_iy_i)}{\sqrt{(\sum{x_i^2})\sum(y_i^2-\frac{(x_iy_i)^2}{x_i^2})}}$ 

= $\frac{\sqrt{(n-1)}\sum(x_iy_i)}{\sqrt{(\sum{x_i^2})(\sum{y_i^2})-(\sum{x_iy_i})^2}}$

In [8]:
n = 100
sum_xi2 = np.sum(x**2)
sum_yi2 = np.sum(y**2)
sum_xiyi = np.sum(x * y)

numerator = np.sqrt(n - 1) * sum_xiyi
denominator = np.sqrt((sum_xi2) * (sum_yi2) - (sum_xiyi)**2)

t = numerator / denominator
t

np.float64(16.898417063035094)

e\)

No matter which variable is regressed on which, the t-statistic for the null hypothesis that the slope is zero will be the same. This is because the t-statistic is a measure of the strength of the relationship between the two variables, and this relationship is symmetric. The formula derived in part (d) shows that the t-statistic depends on the covariance between x and y, as well as their variances, but not on which variable is considered the predictor or response.

In [ ]:
# e)
X = pd.DataFrame({'intercept': np.ones(len(x)),
                  'x': x})
model = sm.OLS(y, X)
resultse1 = model.fit()
summarize(resultse1)

coef  std err       t  P>|t|
intercept -0.0760    0.101  -0.756  0.451
x          1.9686    0.118  16.734  0.000

In [12]:
Y = pd.DataFrame({'intercept': np.ones(len(y)),
                  'y': y})
model = sm.OLS(x, Y)
resultse2 = model.fit()
summarize(resultse2)

coef  std err       t  P>|t|
intercept  0.0095    0.044   0.216  0.829
y          0.3763    0.022  16.734  0.000

The results show that even though the regression is performed with an intercept term, the t-statistic for the slope remains the same regardless of which variable is regressed on which. 